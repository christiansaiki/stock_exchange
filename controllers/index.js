export Health from "./lib/Health";
export RequestHandler from './RequestHandler';
export Exchange from './lib/Exchange';
