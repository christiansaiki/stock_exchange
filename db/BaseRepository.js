
/**
 * BaseRepository - this class has several methods that make the MongoDB calls much easier to handle
 *
 * @export
 * @class BaseRepository
 */
export default class BaseRepository {
	constructor(model) {
		this.model = model;
	}

	async create(info) {
		const entity = new this.model(info);
		return await this.resolve(entity.save());
	}

	async findAndUpdate(query, info, options = { new: true }) {
		return await this.resolve(this.model.findOneAndUpdate(query, info, options));
	}

	async remove(query) {
		return await this.resolve(this.model.remove(query));
	}

	async resolve(promise) {
		let result;
		let error;

		try {
			result = await promise;
		} catch(e) {
			error = e;
		}

		return { error, result };
	}
}
