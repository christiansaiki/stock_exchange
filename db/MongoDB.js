import mongoose from 'mongoose';
import { stocks } from './schemas/stocks_schema';
import helper from '../helper';

/**
 * MongoDB - it is responsible to handle MongoDB connection by using
 * the library called mongoose
 *
 * @export
 * @class MongoDB
 */
export default class MongoDB {
	constructor (env, mongoUrl) {
		this.env = env;
		this.mongoUrl = mongoUrl;
	}

	init () {
		const Schema = mongoose.Schema;
		mongoose.connect(this.mongoUrl);
		helper.logger.info(`MONGO: Connected to ${this.env} - ${this.mongoUrl}`);
		mongoose.set('debug', false);

		this.Stocks = mongoose.model('stocks', stocks(Schema));
	}

	disconnect () {
		mongoose.disconnect();
	}
}
