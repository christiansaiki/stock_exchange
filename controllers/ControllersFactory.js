import { BaseRepository } from "../db";
import {
	Exchange,
	Health
} from ".";

export default class ControllersFactory {
	constructor(mongo) {
		const Stocks = new BaseRepository(mongo.Stocks);
		this.health = new Health;
		this.exchange = new Exchange(Stocks);
	}

	getControllers(url) {
		// Health controller routes
		if(/^(\/health)$/.test(url)) return this.health.check;

		// Exchange controller routes
		if(/^(\/exchange\?.+)$/.test(url)) return this.exchange.buyStock;
		return this.notFound;
	}

	/**
	 * notFound - Only returns 404
	 *
	 * @returns {number}
	 * @memberof ControllersFactory
	 */
	notFound(res) {
		return res.send(404);
	}
}
