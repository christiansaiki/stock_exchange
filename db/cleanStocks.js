import BaseRepository from './BaseRepository';
import MongoDB from './MongoDB';

const env = process.env.NODE_ENV || 'local';
const mongoUrl = process.env.MONGO_URL || 'mongodb://localhost/stock_exchange';

// Init Mongo
const mongo = new MongoDB(env, mongoUrl);
mongo.init();

const Stocks = new BaseRepository(mongo.Stocks);

Stocks.remove({})
	.then(() => {
		console.log('finished');
		process.exit();
	})
	.catch((err) => {
		console.log(err);
		process.exit();
	});
