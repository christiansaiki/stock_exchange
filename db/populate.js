import BaseRepository from './BaseRepository';
import MongoDB from './MongoDB';

const stocks = [{
	company_id: 'C1',
	countries: ['US', 'FR'],
	budget: 1,
	base_bid: 0.1,
	categories: ['Automobile', 'Finance']
}, {
	company_id: 'C2',
	countries: ['IN', 'US'],
	budget: 2,
	base_bid: 0.3,
	categories: ['Finance', 'IT']
}, {
	company_id: 'C3',
	countries: ['US', 'RU'],
	budget: 3,
	base_bid: 0.05,
	categories: ['IT', 'Automobile']
}];

Promise.each = async (arr, fn) => {
	for(const item of arr) await fn(item);
};

module.exports = populateDb;

function populateDb (mongo, shouldDisconnect) {
	const env = process.env.NODE_ENV || 'local';
	const mongoUrl = process.env.MONGO_URL || 'mongodb://localhost/stock_exchange';


	// Init Mongo
	if(!mongo) {
		mongo = new MongoDB(env, mongoUrl);
		mongo.init();
	}
	const Stocks = new BaseRepository(mongo.Stocks);

	Promise.each(stocks, async stock => {
		await Stocks.findAndUpdate({company_id: stock.company_id}, stock, {upsert: true});
	}).then(() => {
		console.log('finished');
		if (shouldDisconnect) {
			console.log('Disconnecting');
			mongo.disconnect();
		}
	})
		.catch((err) => {
			console.log(err);
			if (shouldDisconnect) mongo.disconnect();
		});
};

