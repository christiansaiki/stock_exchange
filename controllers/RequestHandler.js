import ControllersFactory from './ControllersFactory';

export default class RequestHandler {
	constructor(mongo) {
		this.controllersFactory = new ControllersFactory(mongo);

		this.get = this.get.bind(this);
	}

	/**
	 * get - this method handles all the GET requests
	 *
	 * @param {object} req - fastify request object
	 * @param {object} res - fastify response object
	 * @memberof RequestHandler
	 */
	async get(req, res) {
		const { raw, params, query } = req;
		await this.controllersFactory.getControllers(raw.url)(res, params, query);
	}
}
