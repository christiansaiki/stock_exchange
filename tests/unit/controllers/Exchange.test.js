import {Exchange} from '../../../controllers';
import {BaseRepository} from '../../../db';
import sinon from 'sinon';

const exchangeRepository = new BaseRepository();

describe('Exchange Controller - budgetCheck', () => {
	test('budgetCheck - only C1 passes', () => {
		const exchange = new Exchange();
		const result = exchange.budgetCheck([{
			company_id: 'C1',
			budget: 10
		},
		{
			company_id: 'C2',
			budget: 1
		},
		{
			company_id: 'C3',
			budget: 9
		}
		], 10);
		expect(result).toMatchSnapshot();
	});

	test('budgetCheck - everyone passes', () => {
		const exchange = new Exchange();
		const result = exchange.budgetCheck([{
			company_id: 'C1',
			budget: 10
		},
		{
			company_id: 'C2',
			budget: 15
		},
		{
			company_id: 'C3',
			budget: 90
		}
		], 10);
		expect(result).toMatchSnapshot();
	});

	test('budgetCheck - noone passes', () => {
		const exchange = new Exchange();
		const result = exchange.budgetCheck([{
			company_id: 'C1',
			budget: 0
		},
		{
			company_id: 'C2',
			budget: 4
		},
		{
			company_id: 'C3',
			budget: 3
		}
		], 10);
		expect(result).toMatchSnapshot();
	});
});

describe('baseBidCheck', () => {
	test('baseBidCheck - noone passes', () => {
		const exchange = new Exchange();
		const result = exchange.baseBidCheck([{
			company_id: 'C1',
			base_bid: 0.05
		},
		{
			company_id: 'C2',
			base_bid: 0.1
		},
		{
			company_id: 'C3',
			base_bid: 0.15
		}
		], 0.02);
		expect(result).toMatchSnapshot();
	});

	test('baseBidCheck - everyone passes', () => {
		const exchange = new Exchange();
		const result = exchange.baseBidCheck([{
			company_id: 'C1',
			base_bid: 0.05
		},
		{
			company_id: 'C2',
			base_bid: 0.1
		},
		{
			company_id: 'C3',
			base_bid: 0.15
		}
		], 1);
		expect(result).toMatchSnapshot();
	});

	test('baseBidCheck - just C1 passes', () => {
		const exchange = new Exchange();
		const result = exchange.baseBidCheck([{
			company_id: 'C1',
			base_bid: 0.05
		},
		{
			company_id: 'C2',
			base_bid: 0.1
		},
		{
			company_id: 'C3',
			base_bid: 0.15
		}
		], 0.05);
		expect(result).toMatchSnapshot();
	});
});

describe('sortByHighestBid', () => {
	test('sortByHighestBid - C1, C3, C2', () => {
		const exchange = new Exchange();
		const result = exchange.sortByHighestBid([{
			company_id: 'C1',
			base_bid: 0.59
		},
		{
			company_id: 'C2',
			base_bid: 0.14
		},
		{
			company_id: 'C3',
			base_bid: 0.15
		}
		]);
		expect(result).toMatchSnapshot();
	});

	test('sortByHighestBid - C3, C2, C1', () => {
		const exchange = new Exchange();
		const result = exchange.sortByHighestBid([{
			company_id: 'C1',
			base_bid: 0.59
		},
		{
			company_id: 'C2',
			base_bid: 0.64
		},
		{
			company_id: 'C3',
			base_bid: 1
		}
		]);
		expect(result).toMatchSnapshot();
	});

	test('sortByHighestBid - C1, C2, C3', () => {
		const exchange = new Exchange();
		const result = exchange.sortByHighestBid([{
			company_id: 'C1',
			base_bid: 0.59
		},
		{
			company_id: 'C2',
			base_bid: 0.54
		},
		{
			company_id: 'C3',
			base_bid: 0.123
		}
		]);
		expect(result).toMatchSnapshot();
	});
});

describe('buyStock', () => {
	let sandbox, findAllStub, findAndUpdate;
	const exchange = new Exchange(exchangeRepository);
	beforeEach(() => {
		sandbox = sinon.createSandbox();
		findAllStub = sandbox.stub(exchangeRepository, 'findAll');
		findAndUpdate = sandbox.stub(exchangeRepository, 'findAndUpdate');
	});

	afterEach(() => sandbox.restore());

	test('buyStock promise err - error in server during query', async () => {
		findAllStub.returns({ error: true });
		const result = await exchange.buyStock({});
		expect(result).toMatchSnapshot();
	});

	test('buyStock - No Companies Passed from Targeting', async () => {
		findAllStub.returns({ result: [] });
		const result = await exchange.buyStock({});
		expect(result).toMatchSnapshot();
	});

	test('buyStock - No Companies Passed from Budget', async () => {
		findAllStub.returns({
			result: [{
				company_id: 'C1',
				budget: 0.1
			}]
		});
		const result = await exchange.buyStock({BaseBid: 100});
		expect(result).toMatchSnapshot();
	});

	test('buyStock - No Companies Passed from BaseBid check', async () => {
		findAllStub.returns({
			result: [{
				company_id: 'C1',
				budget: 1,
				base_bid: 0.5
			}]
		});
		const result = await exchange.buyStock({BaseBid: 40});
		expect(result).toMatchSnapshot();
	});

	test('buyStock - Winner C1', async () => {
		findAllStub.returns({
			result: [{
				company_id: 'C1',
				budget: 1,
				base_bid: 0.5
			}]
		});
		findAndUpdate.returns({});
		const result = await exchange.buyStock({BaseBid: 50});
		expect(result).toMatchSnapshot();
	});

	test('buyStock - Error in Server while updating', async () => {
		findAllStub.returns({
			result: [{
				company_id: 'C1',
				budget: 1,
				base_bid: 0.5
			}]
		});
		findAndUpdate.returns({error: true});
		const result = await exchange.buyStock({BaseBid: 50});
		expect(result).toMatchSnapshot();
	});
});