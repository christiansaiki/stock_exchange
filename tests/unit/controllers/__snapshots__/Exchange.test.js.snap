// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exchange Controller - budgetCheck budgetCheck - everyone passes 1`] = `
Array [
  Object {
    "budget": 10,
    "company_id": "C1",
  },
  Object {
    "budget": 15,
    "company_id": "C2",
  },
  Object {
    "budget": 90,
    "company_id": "C3",
  },
]
`;

exports[`Exchange Controller - budgetCheck budgetCheck - noone passes 1`] = `Array []`;

exports[`Exchange Controller - budgetCheck budgetCheck - only C1 passes 1`] = `
Array [
  Object {
    "budget": 10,
    "company_id": "C1",
  },
]
`;

exports[`baseBidCheck baseBidCheck - everyone passes 1`] = `
Array [
  Object {
    "base_bid": 0.05,
    "company_id": "C1",
  },
  Object {
    "base_bid": 0.1,
    "company_id": "C2",
  },
  Object {
    "base_bid": 0.15,
    "company_id": "C3",
  },
]
`;

exports[`baseBidCheck baseBidCheck - just C1 passes 1`] = `
Array [
  Object {
    "base_bid": 0.05,
    "company_id": "C1",
  },
]
`;

exports[`baseBidCheck baseBidCheck - noone passes 1`] = `Array []`;

exports[`buyStock buyStock - Error in Server while updating 1`] = `"Error in Server while updating"`;

exports[`buyStock buyStock - No Companies Passed from BaseBid check 1`] = `"No Companies Passed from BaseBid check"`;

exports[`buyStock buyStock - No Companies Passed from Budget 1`] = `"No Companies Passed from Budget"`;

exports[`buyStock buyStock - No Companies Passed from Targeting 1`] = `"No Companies Passed from Targeting"`;

exports[`buyStock buyStock - Winner C1 1`] = `"C1"`;

exports[`buyStock buyStock promise err - error in server during query 1`] = `"Error in Server during query"`;

exports[`sortByHighestBid sortByHighestBid - C1, C2, C3 1`] = `
Array [
  Object {
    "base_bid": 0.59,
    "company_id": "C1",
  },
  Object {
    "base_bid": 0.54,
    "company_id": "C2",
  },
  Object {
    "base_bid": 0.123,
    "company_id": "C3",
  },
]
`;

exports[`sortByHighestBid sortByHighestBid - C1, C3, C2 1`] = `
Array [
  Object {
    "base_bid": 0.59,
    "company_id": "C1",
  },
  Object {
    "base_bid": 0.15,
    "company_id": "C3",
  },
  Object {
    "base_bid": 0.14,
    "company_id": "C2",
  },
]
`;

exports[`sortByHighestBid sortByHighestBid - C3, C2, C1 1`] = `
Array [
  Object {
    "base_bid": 1,
    "company_id": "C3",
  },
  Object {
    "base_bid": 0.64,
    "company_id": "C2",
  },
  Object {
    "base_bid": 0.59,
    "company_id": "C1",
  },
]
`;
